src/ripple/nodestore/impl/DeterministicShard.h

```diff
@@ -0,0 +1,190 @@
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_NODESTORE_DETERMINISTICSHARD_H_INCLUDED
+#define RIPPLE_NODESTORE_DETERMINISTICSHARD_H_INCLUDED
+
+#include <ripple/nodestore/DatabaseShard.h>
+#include <ripple/nodestore/DummyScheduler.h>
+#include <nudb/nudb.hpp>
+#include <set>
+
+// TODO: uncomment after NuDB PR will be approved
+//#define NUDB_CTX_HAS_FLUSH
+
+namespace ripple {
+namespace NodeStore {
+
+/** DeterministicShard class.
+ *
+ * 1. The init() method creates temporary folder tempdir_,
+ *    and the deterministic shard is initialized in that folder.
+ * 2. The store() method adds object to memory pool.
+ * 3. The flush() method stores all objects from memory pool to the shard
+ *    located in tempdir_ in sorted order.
+ * 4. The close(true) method finalizes the shard and moves it from tempdir_
+ *    temporary folder to finaldir_ permanent folder,
+ *    deleting old (non-deterministic) shard located in finaldir_.
+ */
+class DeterministicShard
+{
+    constexpr static std::uint32_t maxMemObjsDef = 16384u;
+    constexpr static std::uint32_t maxMemObjsTest = 16u;
+
+    /* "SHRD" in ASCII */
+    constexpr static std::uint64_t deterministicType = 0x5348524400000000ll;
+
+private:
+    DeterministicShard(DeterministicShard const&) = delete;
+    DeterministicShard&
+    operator=(DeterministicShard const&) = delete;
+
+    /** Creates the object for shard database
+     *
+     * @param app Application object
+     * @param shardDir Directory where shard is located
+     * @param index Index of the shard
+     * @param j Journal to logging
+     */
+    DeterministicShard(
+        Application& app,
+        boost::filesystem::path const& shardDir,
+        std::uint32_t index,
+        beast::Journal j);
+
+    /** Initializes the deterministic shard.
+     *
+     * @param finalKey Serializer of shard's ginal key which consists of:
+     *        shard version (32 bit)
+     *        first ledger sequence in the shard (32 bit)
+     *        last ledger sequence in the shard (32 bit)
+     *        hash of last ledger (256 bits)
+     * @return true is success, false if errored
+     */
+    bool
+    init(Serializer const& finalKey);
+
+public:
+    ~DeterministicShard();
+
+    /** Finalizes, stores and closes the shard.
+     */
+    void
+    close()
+    {
+        close(false);
+    }
+
+    /** Store the object into memory pool
+     *
+     * @param nobj Object to store.
```

Comment:
For shards related code consistency, consider renaming `nObj` to `nodeObject`.

Reply:
Done.